kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -203,11 +203,34 @@ public CloseableIterator<FilteredColumnarBatch> getAddFilesAsColumnarBatches(
   protected Tuple2<Protocol, Metadata> loadTableProtocolAndMetadata(
       Engine engine, Optional<SnapshotHint> snapshotHint, long snapshotVersion) {
 
-    // Exit early if the hint already has the info we need
+    // If the snapshot hint is provided and matches the expected version, load the P&M from version hint.
+    // If not, e.g. for the case of time travel, skip use version hint and fallback to CRC file or log replay.
     if (snapshotHint.isPresent() && snapshotHint.get().getVersion() == snapshotVersion) {
       return new Tuple2<>(snapshotHint.get().getProtocol(), snapshotHint.get().getMetadata());
     }
 
+    // Compute the lower bound for the CRC search
+    // If the snapshot hint is present, we can use it as the lower bound for the CRC search.
+    // TODO: this can be further improved to make the lower bound until the checkpoint version
+    Optional<Long> crcSearchLowerBound = snapshotHint.map(SnapshotHint::getVersion);
+    Optional<CRCInfo> crcInfoOpt =
+        ChecksumReader.getCRCInfo(engine, logSegment.logPath, snapshotVersion, crcSearchLowerBound);
+    if (crcInfoOpt.isPresent()) {
+      CRCInfo crcInfo = crcInfoOpt.get();
+      if (crcInfo.getVersion() == snapshotVersion) {
+        // CRC is related to the desired snapshot version. Load protocol and metadata from CRC.
+        return new Tuple2<>(crcInfo.getProtocol(), crcInfo.getMetadata());
+      }
+      // We found the protocol and metadata in a version older than the one we are looking
+      // for. We need to replay the actions to get the latest protocol and metadata, but
+      // update the hint to read the actions from the version we found to check if the
+      // protocol and metadata are updated in the versions after the one we found.
+      // Building a new snapshotHit for determining the time to end the log replay loop.
+      snapshotHint =
```

Comment:
what if, somehow, the crcInfo version is greater than the target version? we shouldn't set the hint then, since that would be useless. I think we should check for this and log a warning if somehow this happened.

Reply:
@scottsand-db I feel like this brings up for question how we should assign `snapshotHint` (and what kind of query patterns we might expect). As the snapshotHint for a time-travel query might be very different from the one that is useful for the latest snapshot query. So when should we be updating it and to what?

Reply:
From the snapshot manager's perspective, `snapshotHint` is the latest hint for a snapshot for the table, where latest means it has the largest version.

For the LogReplay, I think it is smart to keep the current replay-the-log-until-the-hint code, and to intelligently set the hint to be a value from a CRC, if for example (a) we could not find the CRC for our target version and (b) the CRC we did find is newer (greater) than the hint (and, of course, older (less) than the target version)

Reply:
@huan233usc -- sorry, my comment from above still remains. << what if, somehow, the crcInfo version is greater than the target version? we shouldn't set the hint then, since that would be useless. I think we should check for this and log a warning if somehow this happened. >>

let's add a `io.delta.kernel.internal.util.Preconditions.checkArgument` call that crcInfoOpt, if present, has a version >= crcSearchLowerBound && version <= target`

Reply:
Added a check.  Initially I believe that ChecksumReader.getCRCInfo may help enforce version >= crcSearchLowerBound && version <= target. you made a good point that we should check that before setting the snapshot hint as things may go wrong for arbitrary reason.